src/osd.c

```diff
@@ -223,6 +223,138 @@ preview_cycled_view(struct view *view)
+	set_cairo_color(cairo, theme->osd_bg_color);
+	cairo_rectangle(cairo, 0, 0, w, h);
+	cairo_fill(cairo);
+
+	/* Draw border */
+	set_cairo_color(cairo, theme->osd_border_color);
+	draw_cairo_border(cairo, w, h, theme->osd_border_width);
+
+	/* Highlight current window */
+	if (current_view_index != -1) {
+		int y = OSD_BORDER_WIDTH + OSD_ITEM_HEIGHT * current_view_index;
+
+		if (show_workspace) {
+			/* workspace indicator */
+			y += OSD_ITEM_HEIGHT;
+		}
+
+		set_cairo_color(cairo, theme->osd_label_text_color);
+		cairo_rectangle(cairo, OSD_BORDER_WIDTH, y,
+			OSD_ITEM_WIDTH, OSD_ITEM_HEIGHT);
+		cairo_stroke(cairo);
+	}
+
+	/* Set up text rendering */
+	set_cairo_color(cairo, theme->osd_label_text_color);
+	PangoLayout *layout = pango_cairo_create_layout(cairo);
+	pango_layout_set_width(layout,
+		(OSD_ITEM_WIDTH - 2 * OSD_ITEM_PADDING) * PANGO_SCALE);
+	pango_layout_set_ellipsize(layout, PANGO_ELLIPSIZE_END);
+
+	PangoFontDescription *desc = font_to_pango_desc(&rc.font_osd);
+	pango_layout_set_font_description(layout, desc);
+
+	PangoTabArray *tabs = pango_tab_array_new_with_positions(2, TRUE,
+		PANGO_TAB_LEFT, OSD_TAB1, PANGO_TAB_LEFT, OSD_TAB2);
+	pango_layout_set_tabs(layout, tabs);
+	pango_tab_array_free(tabs);
+
+	pango_cairo_update_layout(cairo, layout);
+
+	int y = OSD_BORDER_WIDTH;
+
+	/* Center text entries on the y axis */
+	int y_offset = (OSD_ITEM_HEIGHT - font_height(&rc.font_osd)) / 2;
+	y += y_offset;
+
+	/* Draw workspace indicator */
+	if (show_workspace) {
+		/* Center workspace indicator on the x axis */
+		int x = font_width(&rc.font_osd, workspace_name);
+		x = (OSD_ITEM_WIDTH - x) / 2;
+		cairo_move_to(cairo, x, y);
+		PangoWeight weight = pango_font_description_get_weight(desc);
+		pango_font_description_set_weight(desc, PANGO_WEIGHT_BOLD);
+		pango_layout_set_font_description(layout, desc);
+		pango_layout_set_text(layout, workspace_name, -1);
+		pango_cairo_show_layout(cairo, layout);
+		pango_font_description_set_weight(desc, weight);
+		pango_layout_set_font_description(layout, desc);
+		y += OSD_ITEM_HEIGHT;
+	}
+	pango_font_description_free(desc);
+
+	struct buf buf;
+	buf_init(&buf);
+
+	/* Draw text for each node */
+	wl_list_for_each_reverse(node, node_list, link) {
+		struct view *view = node_view_from_node(node);
+		if (!isfocusable(view)) {
+			continue;
+		}
+		buf.len = 0;
+		cairo_move_to(cairo, OSD_BORDER_WIDTH + OSD_ITEM_PADDING, y);
+
+		switch (view->type) {
+		case LAB_XDG_SHELL_VIEW:
+			buf_add(&buf, "[xdg-shell]\t");
+			buf_add(&buf, get_formatted_app_id(view));
+			buf_add(&buf, "\t");
+			break;
+#if HAVE_XWAYLAND
+		case LAB_XWAYLAND_VIEW:
+			buf_add(&buf, "[xwayland]\t");
+			buf_add(&buf, view_get_string_prop(view, "class"));
+			buf_add(&buf, "\t");
+			break;
+#endif
+		}
+
+		if (is_title_different(view)) {
+			buf_add(&buf, view_get_string_prop(view, "title"));
+		}
+
+		pango_layout_set_text(layout, buf.buf, -1);
+		pango_cairo_show_layout(cairo, layout);
+		y += OSD_ITEM_HEIGHT;
```

Comment:
between `pango_cairo_show_layout()` and `y += OSD_ITEM_HEIGHT` this can be added:
```c
		if (view == server->osd_state.cycle_view) {
			/* Highlight current window */
			cairo_rectangle(cairo, OSD_BORDER_WIDTH, y - y_offset,
				OSD_ITEM_WIDTH, OSD_ITEM_HEIGHT);
			cairo_stroke(cairo);
		}
```